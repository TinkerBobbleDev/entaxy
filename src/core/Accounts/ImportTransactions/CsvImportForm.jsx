/* eslint-disable no-console */
/* eslint-disable react/no-array-index-key */
import React from 'react'
import { withStyles } from '@material-ui/core/styles'
import PropTypes from 'prop-types'
import Paper from '@material-ui/core/Paper'
import Table from '@material-ui/core/Table'
import TableHead from '@material-ui/core/TableHead'
import TableBody from '@material-ui/core/TableBody'
import TableCell from '@material-ui/core/TableCell'
import TableRow from '@material-ui/core/TableRow'
import MenuItem from '@material-ui/core/MenuItem'
import FormControl from '@material-ui/core/FormControl'
import FormControlLabel from '@material-ui/core/FormControlLabel'
import Checkbox from '@material-ui/core/Checkbox'
import Typography from '@material-ui/core/Typography'
import InputLabel from '@material-ui/core/InputLabel'
import Select from '@material-ui/core/Select'
import Button from '@material-ui/core/Button'
import Divider from '@material-ui/core/Divider'
import CircularProgress from '@material-ui/core/CircularProgress'
import green from '@material-ui/core/colors/green'
import Icon from '@mdi/react'
import { mdiFileDelimited } from '@mdi/js'
import CsvDropzone from './CsvDropzone'
import CsvParser from '../../../store/transactions/CsvParsers/CsvParser'

const styles = (theme) => {
  return ({
    root: {
      display: 'flex',
      justifyContent: 'center',
      margin: '20px'
    },
    fileDetails: {
      display: 'flex',
      padding: 5,
      background: theme.palette.secondary.light,
      color: theme.palette.primary.contrastText,
      borderRadius: 4,
      marginBottom: 10
    },
    csvFileIcon: {
      marginRight: 5,
      verticalAlign: 'text-bottom',
      fill: '#fff'
    },
    formField: {
      width: '100%'
    },
    table: {
      marginBottom: 20
    },
    tableHeader: {
      height: 20,
      background: theme.palette.grey[500]
    },
    tableCell: {
      width: 150,
      color: 'white'
    },
    sample: {
      color: theme.palette.grey[500],
      fontSize: 11,
      width: 150,
      overflow: 'hidden',
      whiteSpace: 'nowrap'
    },
    elipsys: {
      textOverflow: 'ellipsis'
    },
    formOptions: {
      display: 'flex',
      justifyContent: 'center',
      flexDirection: 'column',
      marginTop: 10,
      padding: 10,
      border: 'solid #999 2px'
    },
    formActions: {
      display: 'flex',
      'justify-content': 'flex-end',
      'padding-top': '10px'
    },
    buttonWrapper: {
      margin: theme.spacing(1),
      position: 'relative'
    },
    buttonSuccess: {
      backgroundColor: green[500],
      '&:hover': {
        backgroundColor: green[700]
      }
    },
    buttonProgress: {
      color: green[500],
      position: 'absolute',
      top: '50%',
      left: '50%',
      marginTop: -12,
      marginLeft: -12
    }
  })
}

const DONT_IMPORT = 'Don\'t import'

export class CsvImportFormComponent extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      isSubmitting: false,
      file: null,
      error: null,
      csvHeader: null,
      dateFormat: null
    }
    this.parser = new CsvParser(this.props.budgetRules)
  }

  handleFileUpload = async (acceptedFiles) => {
    if (acceptedFiles[0].type !== 'text/csv') {
      this.setState({
        file: null,
        error: 'The file you uploaded is not a CSV file'
      })
    } else {
      await this.parser.parse(acceptedFiles[0])
      this.setState({
        file: acceptedFiles[0],
        error: null,
        csvHeader: this.parser.csvHeader,
        dateFormat: this.parser.dateFormat
      })
    }
  }

  handleChange = ({ target }) => {
    this.parser.mapColumnToTransactionField({
      columnIndex: parseInt(target.name, 10),
      transactionField: target.value
    })
    this.setState({ csvHeader: this.parser.csvHeader })
  }

  handleChangeDateFormat = ({ target }) => {
    this.parser.dateFormat = target.value
    this.setState({ dateFormat: this.parser.dateFormat })
  }

  handleChangeNoHeaderRow = ({ target }) => {
    this.parser.noHeaderRow = target.checked
    this.setState({ noHeaderRow: this.parser.noHeaderRow })
  }

  handleSubmit = (event) => {
    this.setState({ isSubmitting: true })
    const { handleParsedData } = this.props
    const { transactions, errors } = this.parser.mapToTransactions()
    handleParsedData(transactions, errors)
    this.setState({ isSubmitting: false })
    event.preventDefault()
  }

  render() {
    const {
      classes,
      onCancel,
      account
    } = this.props

    const {
      file,
      error,
      isSubmitting,
      csvHeader,
      dateFormat
    } = this.state

    return (
      <form onSubmit={this.handleSubmit}>
        <div className={classes.root}>
          {(!file || error) && (
            <CsvDropzone
              className={classes.dropzone}
              handleFileUpload={this.handleFileUpload}
              account={account}
              file={file}
              error={error}
            />
          )}
          {file && !error && (
            <div>
              <Typography variant="subtitle2" className={classes.fileDetails}>
                <Icon path={mdiFileDelimited} size={0.7} className={classes.csvFileIcon} />
                {file.name}
              </Typography>
              <Typography variant="caption">
                Choose the transaction field that matches each CSV columns
              </Typography>

              <Table className={classes.table}>
                <TableHead>
                  <TableRow className={classes.tableHeader}>
                    <TableCell className={classes.tableCell} align="right">CSV column</TableCell>
                    <TableCell className={classes.tableCell}>Transaction field</TableCell>
                  </TableRow>
                </TableHead>
                <TableBody>
                  {csvHeader.map(({ label, transactionField, sample }, index) => (
                    <TableRow key={`${index}-${label}`} selected={transactionField !== DONT_IMPORT}>
                      <TableCell align="right">
                        {label}
                        <div className={classes.sample}>
                          <div className={classes.elipsys}>{sample}</div>
                        </div>
                      </TableCell>
                      <TableCell align="left">
                        <FormControl className={classes.formField}>
                          <Select
                            value={transactionField}
                            onChange={this.handleChange}
                            inputProps={{ name: `${index}` }}
                          >
                            <MenuItem key={DONT_IMPORT} value={DONT_IMPORT}>
                              <Typography color="textSecondary">{DONT_IMPORT}</Typography>
                            </MenuItem>
                            {Object.keys(this.parser.transactionFields).map(field => (
                              <MenuItem
                                key={field}
                                value={field}
                                disabled={this.parser.isFieldSelected(field)}
                              >
                                {this.parser.transactionFields[field].label}
                              </MenuItem>
                            ))}
                          </Select>
                        </FormControl>
                      </TableCell>
                    </TableRow>
                  ))}
                </TableBody>
              </Table>
              <Typography variant="subtitle2">Import options</Typography>
              <Paper elevation={0} className={classes.formOptions}>
                <FormControlLabel
                  className={classes.formField}
                  control={(
                    <Checkbox
                      checked={this.state.noHeaderRow}
                      onChange={this.handleChangeNoHeaderRow}
                      value="noHeaderRow"
                    />
                  )}
                  label="No header row"
                />
                <FormControl className={classes.formField}>
                  <InputLabel htmlFor="dateFormat">Date format</InputLabel>
                  <Select
                    value={dateFormat}
                    onChange={this.handleChangeDateFormat}
                    inputProps={{ name: 'dateFormat' }}
                  >
                    {this.parser.dateFormats.map(format => (
                      <MenuItem key={format} value={format}>{format}</MenuItem>
                    ))}
                  </Select>
                </FormControl>
              </Paper>
            </div>
          )}
        </div>
        <Divider />
        <div className={classes.formActions}>
          <div className={classes.buttonWrapper}>
            <Button
              type="submit"
              variant="contained"
              color="primary"
              disabled={isSubmitting || file === null || error !== null}
              onClick={this.handleButtonClick}
            >
              Import
            </Button>
            {isSubmitting && <CircularProgress size={24} className={classes.buttonProgress} />}
          </div>
          <Button onClick={onCancel} color="secondary">Cancel</Button>
        </div>
      </form>
    )
  }
}

CsvImportFormComponent.propTypes = {
  classes: PropTypes.object.isRequired,
  handleParsedData: PropTypes.func.isRequired,
  onCancel: PropTypes.func.isRequired,
  account: PropTypes.object.isRequired,
  budgetRules: PropTypes.object.isRequired
}

export default withStyles(styles)(CsvImportFormComponent)
