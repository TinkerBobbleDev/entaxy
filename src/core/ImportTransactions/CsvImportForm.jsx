import React from 'react'
import { compose } from 'recompose'
import { withStyles } from '@material-ui/core/styles'
import PropTypes from 'prop-types'
import { withFormik } from 'formik'
import Button from '@material-ui/core/Button'
import Divider from '@material-ui/core/Divider'
import CircularProgress from '@material-ui/core/CircularProgress'
import green from '@material-ui/core/colors/green'
import CsvDropzone from './CsvDropzone'
import CsvImportFields from './CsvImportFields'
import RbcCsvParser from '../../store/transactions/CsvParsers/RbcCsvParser'
import BmoCsvParser from '../../store/transactions/CsvParsers/BmoCsvParser'
import TdCsvParser from '../../store/transactions/CsvParsers/TdCsvParser'
import TangerineCsvParser from '../../store/transactions/CsvParsers/TangerineCsvParser'

const styles = theme => ({
  root: {
    display: 'flex',
    'justify-content': 'center'
  },
  input: {
    margin: '5px',
    width: 200
  },
  formActions: {
    display: 'flex',
    'justify-content': 'flex-end',
    'padding-top': '10px'
  },
  wrapper: {
    margin: theme.spacing.unit,
    position: 'relative'
  },
  buttonSuccess: {
    backgroundColor: green[500],
    '&:hover': {
      backgroundColor: green[700]
    }
  },
  buttonProgress: {
    color: green[500],
    position: 'absolute',
    top: '50%',
    left: '50%',
    marginTop: -12,
    marginLeft: -12
  }
})

const parsers = {
  RBC: RbcCsvParser,
  BMO: BmoCsvParser,
  TD: TdCsvParser,
  Tangerine: TangerineCsvParser
}

class CsvImportForm extends React.Component {
  state = {
    file: null,
    error: null
  }

  handleFileUpload = (acceptedFiles) => {
    this.setState({
      file: acceptedFiles[0],
      error: (acceptedFiles[0].type === 'text/csv' ? null : 'The file you uploaded is not a CSV file')
    })
    this.props.setFieldValue('file', acceptedFiles[0])
  }

  render() {
    const {
      classes,
      handleSubmit,
      values,
      handleChange,
      onCancel,
      institution,
      isSubmitting
    } = this.props

    const {
      file,
      error
    } = this.state

    return (
      <form onSubmit={handleSubmit}>
        <div className={classes.root}>
          <CsvImportFields
            institution={institution}
            values={values}
            handleChange={handleChange}
          />
          <CsvDropzone
            className={classes.dropzone}
            handleFileUpload={this.handleFileUpload}
            institution={institution}
            file={file}
            error={error}
          />
        </div>
        <Divider />
        <div className={classes.formActions}>
          <div className={classes.wrapper}>
            <Button
              type="submit"
              variant="contained"
              color="primary"
              disabled={isSubmitting || file === null || error !== null}
              onClick={this.handleButtonClick}
            >
              Import
            </Button>
            {isSubmitting && <CircularProgress size={24} className={classes.buttonProgress} />}
          </div>
          <Button onClick={onCancel} color="secondary">Cancel</Button>
        </div>
      </form>
    )
  }
}

CsvImportForm.propTypes = {
  classes: PropTypes.object.isRequired,
  handleSubmit: PropTypes.func.isRequired,
  values: PropTypes.object.isRequired,
  handleChange: PropTypes.func.isRequired,
  onCancel: PropTypes.func.isRequired,
  institution: PropTypes.string.isRequired,
  setFieldValue: PropTypes.func.isRequired,
  isSubmitting: PropTypes.bool.isRequired
}

export default compose(
  withStyles(styles),
  withFormik({
    mapPropsToValues: ({ institution }) => ({
      institution,
      account: '',
      type: '',
      ticker: '',
      shares: '',
      bookValue: '',
      createdAt: new Date(),
      file: null
    }),
    handleSubmit: (values, { props, setSubmitting }) => {
      setSubmitting(true)
      new parsers[values.institution]().parse(values.file, values)
        .then(({ transactions, errors }) => {
          setSubmitting(false)
          props.handleParsedData(transactions, errors)
        })
    }
  })
)(CsvImportForm)
