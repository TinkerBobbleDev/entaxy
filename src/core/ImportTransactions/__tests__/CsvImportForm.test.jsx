import React from 'react'
import renderer from 'react-test-renderer'
import CsvImportForm from '../CsvImportForm'

describe('CsvImportForm', () => {
  const mockHandleSubmit = jest.fn()
  const mockHandleChange = jest.fn()
  const mockOnCancel = jest.fn()

  const props = {
    classes: { }
  }

  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('matches snapshot for BMO', () => {
    const component = renderer.create((
      <CsvImportForm
        handleSubmit={mockHandleSubmit}
        handleChange={mockHandleChange}
        onCancel={mockOnCancel}
        isSubmitting={false}
        institution="BMO"
        values={{ ticker: 'CAD' }}
        classes={{ ...props }}
      />
    ))
    expect(component.toJSON()).toMatchSnapshot()
  })
})
