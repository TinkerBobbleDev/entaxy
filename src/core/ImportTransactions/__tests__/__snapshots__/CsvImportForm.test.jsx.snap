// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CsvImportForm matches snapshot for BMO 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="WithFormik-CsvImportForm--root-1"
  >
    <div
      className="CsvImportFields-root-7"
    >
      <p
        className="MuiTypography-root-9 MuiTypography-body1-18 MuiTypography-gutterBottom-36"
      >
        Fill out the following field
      </p>
      <div
        className="MuiFormControl-root-44 CsvImportFields-textField-8"
      >
        <label
          className="MuiFormLabel-root-55 MuiInputLabel-root-48 MuiInputLabel-formControl-49 MuiInputLabel-animated-52"
          data-shrink={false}
        >
          Currency
        </label>
        <div
          className="MuiInputBase-root-75 MuiInput-root-62 MuiInput-underline-66 MuiInputBase-formControl-76 MuiInput-formControl-63"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            aria-label="Currency"
            className="MuiInputBase-input-85 MuiInput-input-70"
            disabled={false}
            maxLength={6}
            name="ticker"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={true}
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      aria-disabled={false}
      className="CsvDropzone-root-92"
      onClick={[Function]}
      onDragEnter={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
      style={
        Object {
          "position": "relative",
        }
      }
    >
      <div>
        <div>
          <h4
            className="MuiTypography-root-9 MuiTypography-h4-24 MuiTypography-colorTextSecondary-42 MuiTypography-alignCenter-32"
          >
            Drag a CSV file here
          </h4>
          <h6
            className="MuiTypography-root-9 MuiTypography-subtitle2-28 MuiTypography-colorTextSecondary-42 MuiTypography-alignCenter-32"
          >
            or click to select a file to upload.
          </h6>
        </div>
      </div>
      <input
        autoComplete="off"
        disabled={false}
        multiple={false}
        onChange={[Function]}
        style={
          Object {
            "bottom": 0,
            "left": 0,
            "opacity": 0.00001,
            "pointerEvents": "none",
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
        type="file"
      />
    </div>
  </div>
  <hr
    className="MuiDivider-root-94"
  />
  <div
    className="WithFormik-CsvImportForm--formActions-3"
  >
    <div
      className="WithFormik-CsvImportForm--wrapper-4"
    >
      <button
        className="MuiButtonBase-root-124 MuiButtonBase-disabled-125 MuiButton-root-98 MuiButton-contained-109 MuiButton-containedPrimary-110 MuiButton-raised-112 MuiButton-raisedPrimary-113 MuiButton-disabled-118"
        disabled={true}
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="-1"
        type="submit"
      >
        <span
          className="MuiButton-label-99"
        >
          Import
        </span>
      </button>
    </div>
    <button
      className="MuiButtonBase-root-124 MuiButton-root-98 MuiButton-text-100 MuiButton-textSecondary-102 MuiButton-flat-103 MuiButton-flatSecondary-105"
      disabled={false}
      onBlur={[Function]}
      onClick={[MockFunction]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiButton-label-99"
      >
        Cancel
      </span>
      <span
        className="MuiTouchRipple-root-127"
      />
    </button>
  </div>
</form>
`;
